example/mailto/mailto.cpp

```diff
@@ -0,0 +1,773 @@
+        if (!r)
+            return r.error();
+
+        mailto_view m;
+        m.u_ = *r;
+        if (m.u_.scheme() != "mailto")
+            return urls::grammar::error::invalid;
+        if (m.u_.has_authority())
+            return urls::grammar::error::invalid;
+        if (!is_to(m.u_.encoded_path()))
+            return urls::grammar::error::invalid;
+        for (auto p: m.u_.encoded_params())
+            if (!is_header(p))
+                return urls::grammar::error::invalid;
+
+        return m;
+    }
+
+private:
+    // A long set of functions to validate email
+    // addresses according to RFC6068
+    static
+    bool
+    is_to(urls::string_view s) noexcept
+    {
+        if (s.empty())
+            return true;
+        auto p = s.find(',');
+        while (p != urls::string_view::npos)
+        {
+            if (!is_address(s.substr(0, p)))
+                return false;
+            s = s.substr(p + 1);
+            p = s.find(',');
+        }
+        if (!is_address(s))
+            return false;
+        return true;
+    }
+
+    static
+    bool
+    is_address(urls::string_view s) noexcept
+    {
+        // addr-spec = local-part "@" domain
+        std::size_t p = s.find('@');
+        if (p == urls::string_view::npos)
+            return false;
+        return is_local_part(s.substr(0, p)) &&
+               is_domain(s.substr(p + 1));
+    }
+
+    static
+    bool
+    is_local_part(urls::string_view s) noexcept
+    {
+        // local-part = dot-atom-text / quoted-string
+        return is_dot_atom_text(s) ||
+               is_quoted_string(s);
+    }
+
+    static
+    bool
+    is_dot_atom_text(urls::string_view s) noexcept
+    {
+        // dot-atom-text = 1*atext *("." 1*atext)
+        if (s.empty())
+            return false;
+        bool prev_dot = true;
+        for (std::size_t i = 0; i < s.size(); ++i)
+        {
+            if (prev_dot)
+            {
+                if (is_atext(s[i]))
+                    prev_dot = false;
+                else
+                    return false;
+            }
+            else if (!is_atext(s[i]))
+            {
+                if (s[i] == '.')
+                    prev_dot = true;
+                else
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    static
+    bool
+    is_atext(char c) noexcept
+    {
+        // atext = ALPHA / DIGIT / "!" / "#" / "$" / "%" / "&" / "'" /
+        // "*" / "+" / "-" / "/" / "=" / "?" / "^" / "_" / "`" / "{" /
+        // "|" / "}" / "~"
+        return urls::grammar::alnum_chars(c) ||
+               c == '!'  ||
+               (c >= '#' && c <= '\'') ||
+               c == '*'  || c == '+'  ||
+               c == '-'  || c == '/'  ||
+               c == '='  || c == '?'  ||
+               (c >= '^' && c <= '`') ||
+               (c >= '{' && c <= '~');
+    }
+
+    static
+    bool
+    is_quoted_string(urls::string_view s) noexcept
```

Comment:
shouldn't this be expressed as a grammar rule? e.g. `quoted_string_rule`? Even if it is in a `detail` namespace, it looks like you have reimplemented a parser

Reply:
Same as https://github.com/boostorg/url/pull/573#discussion_r982969290